File: services/notification-service/src/main/java/com/visor/school/notification/model/Notification.java
package com.visor.school.notification.model;

import jakarta.persistence.*;
import java.time.Instant;
import java.util.Map;
import java.util.UUID;

/**
 * Notification entity for important events
 */
@Entity
@Table(name = "notifications", indexes = {
    @Index(name = "idx_notification_user", columnList = "user_id"),
    @Index(name = "idx_notification_read", columnList = "user_id,read"),
    @Index(name = "idx_notification_created", columnList = "created_at"),
    @Index(name = "idx_notification_type", columnList = "type")
})
public class Notification {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "user_id", nullable = false)
    private UUID userId;

    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false)
    private NotificationType type;

    @Column(name = "title", nullable = false)
    private String title;

    @Column(name = "message", nullable = false, columnDefinition = "TEXT")
    private String message;

    @Column(name = "metadata", columnDefinition = "JSONB")
    @Convert(converter = JsonMapConverter.class)
    private Map<String, Object> metadata;

    @Column(name = "read", nullable = false)
    private boolean read;

    @Column(name = "read_at")
    private Instant readAt;

    @Column(name = "created_at", nullable = false)
    private Instant createdAt;

    // Default constructor for JPA
    public Notification() {
        this.id = UUID.randomUUID();
        this.createdAt = Instant.now();
        this.read = false;
    }

    public Notification(UUID userId, NotificationType type, String title, String message, Map<String, Object> metadata) {
        this();
        if (title == null || title.isBlank()) {
            throw new IllegalArgumentException("Title cannot be blank");
        }
        if (message == null || message.isBlank()) {
            throw new IllegalArgumentException("Message cannot be blank");
        }
        this.userId = userId;
        this.type = type;
        this.title = title;
        this.message = message;
        this.metadata = metadata;
    }

    // Getters and Setters

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public UUID getUserId() {
        return userId;
    }

    public void setUserId(UUID userId) {
        this.userId = userId;
    }

    public NotificationType getType() {
        return type;
    }

    public void setType(NotificationType type) {
        this.type = type;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    public Map<String, Object> getMetadata() {
        return metadata;
    }

    public void setMetadata(Map<String, Object> metadata) {
        this.metadata = metadata;
    }

    public boolean isRead() {
        return read;
    }

    public void setRead(boolean read) {
        this.read = read;
    }

    public Instant getReadAt() {
        return readAt;
    }

    public void setReadAt(Instant readAt) {
        this.readAt = readAt;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Instant createdAt) {
        this.createdAt = createdAt;
    }
}

File: services/notification-service/src/main/kotlin/com/visor/school/notification/model/Notification.kt
package com.visor.school.notification.model

import jakarta.persistence.*
import java.time.Instant
import java.util.UUID

/**
 * Notification entity for important events
 */
@Entity
@Table(name = "notifications", indexes = [
    Index(name = "idx_notification_user", columnList = "user_id"),
    Index(name = "idx_notification_read", columnList = "user_id,read"),
    Index(name = "idx_notification_created", columnList = "created_at"),
    Index(name = "idx_notification_type", columnList = "type")
])
class Notification(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    val id: UUID = UUID.randomUUID(),

    @Column(name = "user_id", nullable = false)
    val userId: UUID,

    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false)
    val type: NotificationType,

    @Column(name = "title", nullable = false)
    val title: String,

    @Column(name = "message", nullable = false, columnDefinition = "TEXT")
    val message: String,

    @Column(name = "metadata", columnDefinition = "JSONB")
    @Convert(converter = JsonMapConverter::class)
    val metadata: Map<String, Any>? = null,

    @Column(name = "read", nullable = false)
    var read: Boolean = false,

    @Column(name = "read_at")
    var readAt: Instant? = null,

    @Column(name = "created_at", nullable = false)
    val createdAt: Instant = Instant.now()
) {
    init {
        require(title.isNotBlank()) {
            "Title cannot be blank"
        }
        require(message.isNotBlank()) {
            "Message cannot be blank"
        }
    }
}

/**
 * Converter for JSON map storage
 */
@Converter
class JsonMapConverter : jakarta.persistence.AttributeConverter<Map<String, Any>, String> {
    override fun convertToDatabaseColumn(attribute: Map<String, Any>?): String? {
        if (attribute == null) return null
        return com.fasterxml.jackson.databind.ObjectMapper().writeValueAsString(attribute)
    }

    override fun convertToEntityAttribute(dbData: String?): Map<String, Any>? {
        if (dbData == null) return null
        return com.fasterxml.jackson.databind.ObjectMapper().readValue(dbData, Map::class.java) as? Map<String, Any>
    }
}

